slasher/db/kv/indexed_attestations.go

```diff
@@ -105,6 +105,95 @@ func (db *Store) IdxAttsForTarget(ctx context.Context, targetEpoch uint64) ([]*e
 	return idxAtts, err
 }
 
+// IndexedAttestationsForEpoch retrieves all indexed attestations by target epoch
+// by looking into a bucket of attestation roots by epoch first. Then, it
+// retrieves all attestations that match that list of roots.
+func (db *Store) IndexedAttestationsForEpoch(ctx context.Context, targetEpoch uint64) ([]*ethpb.IndexedAttestation, error) {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.IndexedAttestationsForEpoch")
+	defer span.End()
+	var idxAtts []*ethpb.IndexedAttestation
+	key := bytesutil.Bytes8(targetEpoch)
+	err := db.view(func(tx *bolt.Tx) error {
+		rootsBkt := tx.Bucket(indexedAttestationsRootsByTargetBucket)
+		attRoots := rootsBkt.Get(key)
+		splitRoots := make([][]byte, 0)
+		for i := 0; i < len(attRoots); i += 32 {
+			splitRoots = append(splitRoots, attRoots[i:i+32])
+		}
+		attsBkt := tx.Bucket(indexedAttestationsBucket)
+		for i := 0; i < len(splitRoots); i++ {
+			enc := attsBkt.Get(splitRoots[i])
+			idxAtt, err := unmarshalIdxAtt(ctx, enc)
+			if err != nil {
+				return err
+			}
+			idxAtts = append(idxAtts, idxAtt)
+		}
+		return nil
+	})
+	return idxAtts, err
+}
+
+// SaveIncomingIndexedAttestationsForEpoch stores a list of indexed attestations into the db
+// by storing their roots into a bucket using their target epoch as the key.
+func (db *Store) SaveIncomingIndexedAttestationsForEpoch(ctx context.Context, atts []*ethpb.IndexedAttestation) error {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.SaveIndexedAttestations")
+	defer span.End()
+	encoded := make([][]byte, len(atts))
+	encodedRoots := make([][]byte, len(atts))
+	for i := 0; i < len(encoded); i++ {
+		enc, err := proto.Marshal(atts[i])
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal")
+		}
+		encoded[i] = enc
+		root := hashutil.Hash(enc)
+		encodedRoots[i] = root[:]
+	}
+	return db.update(func(tx *bolt.Tx) error {
+		rootsBkt := tx.Bucket(indexedAttestationsRootsByTargetBucket)
+		attsBkt := tx.Bucket(indexedAttestationsBucket)
+
+		for i := 0; i < len(atts); i++ {
```

Comment:
this loop seems expensive why not iterate outside the db update and then only append it once with everything batched together

Reply:
There's no guarantee that all the attestations received are for the same target epoch.

Reply:
this will be expensive to save but we can leave it as it is, maybe we can make a note here that it is inefficient to save it this way.

Reply:
Left a comment. Thanks!